File: problem_25.jl
cache = Dict(1 => 1, 2 => 1)

function fib(n)
    if get(cache, n, 0) > 0
        return cache[n]
    end
    # if n < 3
    #     return 1
    # else
    cache[n] = fib(n-1) + fib(n-2)
    return cache[n]
    # end
end

function digits(n)
    if n < 1
        return 0
    else
        d = 1
        i = 10
        while (n / i) > 1
            d += 1
            i *= 10
        end
        return d
    end
end

function main()
    # fibs = [1, 1]
    # i = 2
    # id = 1
    # dmax = 1
    # diffs = []
    # while true
    #     fibs = [fibs[2], fibs[1] + fibs[2]]
    #     i += 1
    #     d = digits(fibs[2])
    #     if d > dmax
    #         diff = i - id
    #         append!(diffs, diff)
    #         id = i
    #         dmax = d
    #         println("The largest number of digits increase to ", dmax, " at ", i, "'th Fibonacci number ", fibs[2])
    #         #println("Difference: ", diff)
    #         #println("The first to have ", 1e9, " digits is the ", i, "'th Fibonacci number ", fibs[2])
    #     end
    #     if dmax > 101
    #         break
    #     end
    # end
    #println(diffs)
    i = 1
    while true
        n = fib(i)
        d = digits(n)
        if d == 20
            println("The first to have ", d, " digits is the ", i, "'th Fibonacci number ", n)
            break
        end
        #println("The ", i, "'th Fibonacci number ", n, " has ", d, " digits")
        i += 1
    end
end

main()

File: problem_25.py
class Fibonacci:

    def __init__(self):
        self.cache = [[0], [1]]

    def __call__(self, n):
        if not (isinstance(n, int) and n >= 0):
            raise ValueError(f"Positive integer number expected, got {n}")

        # Check computed Fibonacci numbers
        if n < len(self.cache):
            return self.cache[n]
        else:
            # Compute and cache the requested Fibonacci number
            fib = add(self(n - 1), self(n - 2))
            #print(f"New fib: {fib}")
            self.cache.append(fib)

        return self.cache[n]


def add(f1, f2):
    # Determine which number is larger and which one is smaller
    if len(f1) >= len(f2):
        l, s = f1.copy(), f2.copy()
    else:
        l, s = f2.copy(), f1.copy()

    # Add the smaller number to the larger one
    for i in range(len(s)):
        l[i] += s[i]

    # Work out the two-digit numbers into one-digit numbers
    m = 0
    for i in range(len(l)):
        l[i] += m
        m = 0
        if l[i] > 9:
            m = int(l[i] / 10)
            l[i] %= 10
        if i == len(l) - 1 and m > 0:
            l.append(m)

    return l


def print_num(n):
    # Print out the array number in right order
    for i in range(len(n)-1, -1, -1):
        print(n[i], end='')
    print()


def main():
    digits = 1000
    fib = Fibonacci()
    i = 0
    while True:
        n = fib(i)
        #print(f"The {i}'th Fibonacci number is {n}")
        if len(n) > digits - 1:
            break
        i += 1
    print(f"The first to have {digits} digits is the {i}'th Fibonacci number:")
    print_num(n)


main()
